src/player_display.cpp

```diff
@@ -96,6 +96,31 @@ static std::vector<std::pair<bodypart_id, bool>> list_and_combine_bps( const Cha
     return bps;
 }
 
+static std::pair<size_t, size_t> subindex_around_cursor(
+    const size_t list_lenght,
```

Comment:
list_lenght -> list_length. And I believe `calcStartPos` from `output.h` does exactly this?

Reply:
It's a bit different, it does calculate `slice_start`, but not `slice_end`.
But maybe it could be used to reduce code duplicity a bit...

Reply:
I don't yet get `get_option<bool>( "MENU_SCROLL" )` in `calcStartPos` thought...

Reply:
`options.cpp:1659`:

```c++
    add( "MENU_SCROLL", "interface", to_translation( "Centered menu scrolling" ),
         to_translation( "If true, menus will start scrolling in the center of the list, and keep the list centered." ),
         true
       );
```

Reply:
Ok, so this is the existing function:
```cpp
void calcStartPos( int &iStartPos, const int iCurrentLine, const int iContentHeight,
                   const int iNumEntries )
{
    if( iNumEntries <= iContentHeight ) {
        iStartPos = 0;
    } else if( get_option<bool>( "MENU_SCROLL" ) ) {
        iStartPos = iCurrentLine - ( iContentHeight - 1 ) / 2;
        if( iStartPos < 0 ) {
            iStartPos = 0;
        } else if( iStartPos + iContentHeight > iNumEntries ) {
            iStartPos = iNumEntries - iContentHeight;
        }
    } else {
        if( iCurrentLine < iStartPos ) {
            iStartPos = iCurrentLine;
        } else if( iCurrentLine >= iStartPos + iContentHeight ) {
            iStartPos = 1 + iCurrentLine - iContentHeight;
        }
    }
}
```

I believe I'm doing the same it can provide for me on these two lines:
```cpp
    size_t slice_start = std::max( 0, int( cursor_pos ) - int( available_space ) / 2 );
    slice_start = std::min( slice_start, list_lenght - available_space );
```
in this function:
```cpp
static std::pair<size_t, size_t> subindex_around_cursor(
    const size_t list_lenght,
    const size_t available_space,
    const size_t cursor_pos,
    const bool focused
)
/**
 * Return indexes [start, end) that should be displayed from list long `list_lenght`,
 * given that cursor is at position `cursor_pos` and we have `available_space` spaces.
 *
 * Example:
 * list_lenght = 6, available_space = 3, cursor_pos = 2, focused = true;
 * so choose 3 from indexes [0, 1, 2, 3, 4, 5]
 * return {1, 4}
 */
{
    if( !focused || list_lenght <= available_space ) {
        return std::make_pair( 0, std::min( available_space, list_lenght ) );
    }
    size_t slice_start = std::max( 0, int( cursor_pos ) - int( available_space ) / 2 );
    slice_start = std::min( slice_start, list_lenght - available_space );
    size_t slice_end = slice_start + available_space;
    return std::make_pair( slice_start, slice_end );
}
```

The existing `calcStartPos` seems very wasteful to me...

Reply:
Using `calcStartPos` respects the "MENU_SCROLL" option though. That option can be true or false depending on user settings.

Reply:
Ok, I slept and now I understand the simple description of `"MENU_SCROLL"`, will do :)

Reply:
Actually, it's not as easy as plugging it in... it doesn't work while going up while scrolling (it goes up while cursor is on the last line)... But that's fixable, it just needs memory.

Reply:
However, is every menu drawn like this? It looks stupid to me, you should just have a method or a function to which you say "I have space there, and I have X items, figure out scrolling, ask me when you need to draw Xth line, I will provide text".
It seems that every time you make a window, you actually have to make it from characters. And there are lot's of menus in this game!

Or am I missing something?

Reply:
I got that impression from this, autopickup & crafting menus.

Reply:
Having a generic scrollbar class like you describe would be nice, but currently there isn't such a thing.

Reply:
If somebody is reading this @ ZhilkinSerg directed me to src/ui.h which contains `class uilist` described as `uilist: scrolling vertical list menu`. Looks intimidating on my first issue, I made a feature request (#50928) about the "MENU_SCROLL". Will not fix in this PR.